rclrs_examples/src/minimal_publisher.rs

```diff
@@ -22,7 +22,7 @@ fn main() -> Result<(), Error> {
     while context.ok()? {
         message.data = format!("Hello, world! {}", publish_count);
         println!("Publishing: [{}]", message.data);
-        publisher.publish(&message)?;
+        publisher.publish(message.clone())?;
```

Comment:
Would it be possible to implement the `Message` trait for references as well? `rclcpp` does not require developers to create a new copy of a message before publishing it, and it'd be cool if Rust developers had the same experience.

Reply:
Maybe. But one thing I like about having it implemented only for owned values is that it encourages a more efficient coding style by making the clones explicit.

Reply:
However, that's a performance problem. If a user has a large point cloud or an image, cloning it makes little sense, hence why `rclcpp` uses pointers and not instances. And one of the benefits of Rust is the borrow checker, you can guarantee that references don't go out of scope, so you don't have to waste memory.

Reply:
So, there's two different things we could do:
1. Implement `Message` for `&IdiomaticType`
2. Implement `Message` for `&RmwType`

1 is less efficient than the existing impl for an owned `IdiomaticType`. For instance, for an array, when the source array is owned, you can use `Array::map`, but when it's borrowed, you need to collect into a temporary `Vec` and call `try_into()` to create an array again. Bounded strings and sequences would need to be cloned from a borrowed message, instead of moved.

2  could be equally efficient as the existing impl for an owned `RmwType`, i.e. a no-op, at least I think it should be possible with `Cow` or something.

I mostly see the value for 2. Do you also want 1?

Reply:
I'd like to avoid copies anywhere in the code, otherwise we'll have to readd the first item in README that says that messages are copied.

Reply:
Also, idiomatic types in `rclcpp` are passed as shared pointers, not as copies, so giving a worse experience to Rust developers would be unfair when it's technically possible.

Reply:
So you're saying there should be a `Message` impl for `&IdiomaticType` too? I can do that.

Reply:
The only problem is that this will require s lifetime parameter on the trait, making it less ergonomic. But I'll try it out.

Reply:
The only ergonomic way I've found so far is to split up the `Message` trait:

```rust
pub trait Message {
    /// The corresponding RMW-compatible message type.
    type RmwMsg: RmwMessage;

    /// Converts the idiomatic message into a RMW-compatible message.
    fn into_rmw_message(self) -> Self::RmwMsg;
}

pub trait OwnedMessage: Message {
    /// Converts the RMW-compatible message into an idiomatic message.
    fn from_rmw_message(msg: Self::RmwMsg) -> Self;
}
``` 
`Message` is implemented by owned values and references, and `OwnedMessage` only by owned values. `Message` is used on the publishing side, `OwnedMessage` on the subscription side.

Does this architecture sound alright, and do you have suggestions for better names?

Reply:
> So you're saying there should be a Message impl for &IdiomaticType too? I can do that.

Yes, that's what I meant by implementing `Message` for references.

> Does this architecture sound alright, and do you have suggestions for better names?

A priori this look good :+1:

Reply:
Reworked it again and pushed it. Please see the last two commits.

Reply:
I'm resolving this thread since the approach of implementing `Message` for `&T` has been superseded by the `publish()` function taking a `T` and `&T` where `T: Message`, as discussed in this thread: https://github.com/ros2-rust/ros2_rust/pull/80#discussion_r824689568